File: android/app/src/main/java/app/notomorrow/feature/progress/LiftRecords.kt
package app.notomorrow.feature.progress

import androidx.compose.runtime.Immutable
import app.notomorrow.data.relation.CompletedSetRow

/**
 * A lift's all-time records — port of `LiftRecords.swift`, read off the same working sets
 * [LiftSummary] holds (so the numbers match the lift's progress page): the best Epley e1RM
 * (`RecordService.bestSet`'s rule), the heaviest set ([LiftSummary.heaviest]) and the set with
 * the most weight × reps.
 */
@Immutable
data class LiftRecords(
    /** The set behind the best e1RM; its e1RM is [bestE1RMKg], the lift's `current`. */
    val bestE1RM: CompletedSetRow,
    val bestE1RMKg: Double,
    val heaviest: CompletedSetRow,
    val bestVolume: CompletedSetRow,
) {
    companion object {
        /** `null` without a set that counts (weight and reps above zero). */
        fun of(sets: List<CompletedSetRow>): LiftRecords? {
            val counted = sets.filter { it.weightKg > 0 && it.reps > 0 }
            val e1RM = bestE1RM(counted) ?: return null
            return LiftRecords(
                bestE1RM = e1RM,
                bestE1RMKg = e1RM.estimatedOneRepMax,
                heaviest = heaviest(counted) ?: return null,
                bestVolume = bestVolume(counted) ?: return null,
            )
        }

        /** Highest e1RM (ties → heavier). */
        fun bestE1RM(sets: List<CompletedSetRow>): CompletedSetRow? =
            sets.maxWithOrNull(compareBy({ it.estimatedOneRepMax }, { it.weightKg }))

        /** Heaviest weight (ties → more reps). */
        fun heaviest(sets: List<CompletedSetRow>): CompletedSetRow? =
            sets.maxWithOrNull(compareBy({ it.weightKg }, { it.reps }))

        /** Most weight × reps (ties → heavier). */
        fun bestVolume(sets: List<CompletedSetRow>): CompletedSetRow? =
            sets.maxWithOrNull(compareBy({ it.weightKg * it.reps }, { it.weightKg }))

        /**
         * The lifts with a PR, in the order the Records screen lists them: most recent PR first,
         * the order [ProgressDerivations.buildLifts] already sorts them in.
         */
        fun withPRs(lifts: List<LiftSummary>): List<Pair<LiftSummary, LiftRecords>> =
            lifts.mapNotNull { lift ->
                if (lift.lastPR == null) return@mapNotNull null
                of(lift.sets)?.let { lift to it }
            }
    }
}

File: NoTomorrow/Features/Progress/LiftRecords.swift
import Foundation
import SwiftData

/// A lift's all-time records, read off the same working sets `LiftSummary` holds (so the numbers match the lift's
/// progress page): the best Epley e1RM (`RecordService.bestSet`'s rule), the heaviest set (`LiftSummary.heaviest`)
/// and the set with the most weight × reps.
struct LiftRecords: Equatable {
    /// The set behind the best e1RM; its e1RM is `bestE1RMKg`, the lift's `current`.
    var bestE1RM: LiftSet
    var bestE1RMKg: Double
    var heaviest: LiftSet
    var bestVolume: LiftSet

    /// Nil without a set that counts (weight and reps above zero).
    init?(sets: [LiftSet]) {
        let counted = sets.filter { $0.weightKg > 0 && $0.reps > 0 }
        guard let e1RM = Self.bestE1RM(in: counted),
              let heaviest = Self.heaviest(in: counted),
              let volume = Self.bestVolume(in: counted) else { return nil }
        self.bestE1RM = e1RM
        self.bestE1RMKg = e1RM.estimatedOneRepMax
        self.heaviest = heaviest
        self.bestVolume = volume
    }

    /// Highest e1RM (ties → heavier).
    static func bestE1RM(in sets: [LiftSet]) -> LiftSet? {
        sets.max { lhs, rhs in
            if lhs.estimatedOneRepMax != rhs.estimatedOneRepMax { return lhs.estimatedOneRepMax < rhs.estimatedOneRepMax }
            return lhs.weightKg < rhs.weightKg
        }
    }

    /// Heaviest weight (ties → more reps).
    static func heaviest(in sets: [LiftSet]) -> LiftSet? {
        sets.max { lhs, rhs in
            if lhs.weightKg != rhs.weightKg { return lhs.weightKg < rhs.weightKg }
            return lhs.reps < rhs.reps
        }
    }

    /// Most weight × reps (ties → heavier).
    static func bestVolume(in sets: [LiftSet]) -> LiftSet? {
        sets.max { lhs, rhs in
            if lhs.volumeKg != rhs.volumeKg { return lhs.volumeKg < rhs.volumeKg }
            return lhs.weightKg < rhs.weightKg
        }
    }

    /// The lifts with a PR, in the order the Records screen lists them: most recent PR first, the order
    /// `ProgressModel` already sorts `lifts` in.
    static func withPRs(_ lifts: [LiftSummary]) -> [LiftRecordEntry] {
        lifts.compactMap { lift -> LiftRecordEntry? in
            guard lift.lastPR != nil, let records = LiftRecords(sets: lift.sets) else { return nil }
            return LiftRecordEntry(lift: lift, records: records)
        }
    }
}

/// One row of the Records screen.
struct LiftRecordEntry: Identifiable {
    let lift: LiftSummary
    let records: LiftRecords
    var id: PersistentIdentifier { lift.id }
}

extension LiftSet {
    /// Epley, as `SetEntry.estimatedOneRepMax`.
    var estimatedOneRepMax: Double {
        guard reps > 0, weightKg > 0 else { return 0 }
        if reps == 1 { return weightKg }
        return weightKg * (1 + Double(reps) / 30)
    }

    var volumeKg: Double { weightKg * Double(reps) }
}
